packages/grid/x-data-grid/src/DataGrid/useDataGridProps.ts

```diff
@@ -73,39 +74,43 @@ export const DATA_GRID_PROPS_DEFAULT_VALUES: DataGridPropsWithDefaultValues = {
 };
 
 export const useDataGridProps = <R extends GridValidRowModel>(inProps: DataGridProps<R>) => {
-  const themedProps = useThemeProps({ props: inProps, name: 'MuiDataGrid' });
+  const { componentsProps, ...themedProps } = useThemeProps({
+    props: inProps,
+    name: 'MuiDataGrid',
+  });
 
   const localeText = React.useMemo(
     () => ({ ...GRID_DEFAULT_LOCALE_TEXT, ...themedProps.localeText }),
     [themedProps.localeText],
   );
 
-  const components = React.useMemo<GridSlotsComponent>(() => {
-    const overrides = themedProps.components;
+  const slots = React.useMemo<UncapitalizedGridSlotsComponent>(() => {
+    const uncapitalizedDefaultSlots = uncapitalizeObjectKeys(DATA_GRID_DEFAULT_SLOTS_COMPONENTS)!;
+    const overrides =
+      themedProps.slots ??
+      (themedProps.components ? uncapitalizeObjectKeys(themedProps.components) : null);
 
     if (!overrides) {
-      return { ...DATA_GRID_DEFAULT_SLOTS_COMPONENTS };
+      return { ...uncapitalizedDefaultSlots };
     }
 
-    const mergedComponents = {} as GridSlotsComponent;
-    type GridSlots = keyof GridSlotsComponent;
-
-    Object.entries(DATA_GRID_DEFAULT_SLOTS_COMPONENTS).forEach(([key, defaultComponent]) => {
-      mergedComponents[key as GridSlots] =
-        overrides[key as GridSlots] === undefined ? defaultComponent : overrides[key as GridSlots];
-    });
-
-    return mergedComponents;
-  }, [themedProps.components]);
+    type GridSlot = keyof UncapitalizedGridSlotsComponent;
+    return Object.entries(uncapitalizedDefaultSlots).reduce((acc, [key, defaultComponent]) => {
+      const override = overrides[key as GridSlot];
+      const component = override !== undefined ? override : defaultComponent;
```

Comment:
Should we disallow `null` values for slot overrides? 🤔 
Because technically users should not pass `null` values here or some default functionality of Grid may stop working. This is enforced by types already, maybe it's worth it to enforce it in logic as well.

I can't think of some use-cases which may help user by allowing slot override `null`.